### PR TITLE
Softlock/screenlock distinct

### PR DESCRIPTION
**Pull request by johanw666**
Gives the user a choice between setting a password or the Android system lock to lock the app. Is also usefull for Android 4.4 because there the system lock does not work (see https://github.com/signalapp/Signal-Android/issues/8603).

### CODE REVIEW

File: src/org/thoughtcrime/securesms/util/TextSecurePreferences.java [hunk 1]
```diff
@@ -183,6 +183,16 @@
 
   private static final String MEDIA_KEYBOARD_MODE = "pref_media_keyboard_mode";
 
+  public static final String PROTECTION_METHOD_PREF = "pref_signal_protection_method";
+
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: leading whitespace

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 2]
```diff
@@ -315,6 +340,68 @@ public boolean onPreferenceChange(final Preference preference, Object newValue)
...
+      return true;
+    }
+  }
+
+  private class EnableScreenLockToggleListener implements Preference.OnPreferenceChangeListener {
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+      boolean screenlockEnabled = (boolean) newValue;
+
+      if (Build.VERSION.SDK_INT >= 21) {
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: use Build.VERSION_CODES.LOLLIPOP

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 3]
```diff
@@ -53,6 +56,8 @@ public void onAttach(Activity activity) {
   public void onCreate(Bundle paramBundle) {
     super.onCreate(paramBundle);
 
+    protectionMethod = (CheckBoxPreference) this.findPreference(TextSecurePreferences.PROTECTION_METHOD_PREF);
+    enableScreenLock = (CheckBoxPreference) this.findPreference(TextSecurePreferences.SCREEN_LOCK);
     disablePassphrase = (CheckBoxPreference) this.findPreference("pref_enable_passphrase_temporary");
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: you can define "pref_enable_passphrase_temporary" as constant as well, even if you don't use getters/setters for that preference

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 4]
```diff
@@ -107,7 +121,15 @@ private void initializeScreenLockTimeoutSummary() {
   }
 
   private void initializeVisibility() {
-    if (TextSecurePreferences.isPasswordDisabled(getContext())) {
+    // JW:
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: stale comment

File: src/org/thoughtcrime/securesms/util/TextSecurePreferences.java [hunk 5]
```diff
@@ -183,6 +183,16 @@
 
   private static final String MEDIA_KEYBOARD_MODE = "pref_media_keyboard_mode";
 
+  public static final String PROTECTION_METHOD_PREF = "pref_signal_protection_method";
+  
+  public static boolean isProtectionMethodPassphrase(@NonNull Context context) {
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: move these two new functions to the botton of the class to reduce merging conflicts

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 6]
```diff
@@ -87,6 +94,13 @@ public void onResume() {
     else                                                         initializeScreenLockTimeoutSummary();
 
     disablePassphrase.setChecked(!TextSecurePreferences.isPasswordDisabled(getActivity()));
+    // Set the choice between passphrase and Android screenlock. Because this preference
+    // may be undefined when this app is first ran we also check if there is a passphrase
+    // defined, if so, we assume passphrase protection:
+    boolean selectPassphrase = TextSecurePreferences.isProtectionMethodPassphrase(getContext()) ||
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: this should be a function, the same if-statement is repeated later on

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 7]
```diff
@@ -315,6 +340,68 @@ public boolean onPreferenceChange(final Preference preference, Object newValue)
...
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+      boolean usePassphraseProtection = (boolean) newValue;
+
+      TextSecurePreferences.setProtectionMethod(getContext(), usePassphraseProtection);
+
+      // After togggle, we disable both passphrase and Android keylock.
+      // Remove the passphrase if there is one set
+      if (TextSecurePreferences.getBooleanPreference(getContext(), "pref_enable_passphrase_temporary", false))
+      {
+        MasterSecretUtil.changeMasterSecretPassphrase(getActivity(),
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: create a new function to disable the passhprase, both DisablePassphraseClickListener and ProtectionMethodToggleListener use the same code:
```java
        MasterSecretUtil.changeMasterSecretPassphrase(getActivity(),
          KeyCachingService.getMasterSecret(getContext()),
          MasterSecretUtil.UNENCRYPTED_PASSPHRASE);
        // setChecked(false) only in DisablePassphraseClickListener
        // can be moved to the top, before calling the proposed disablePasshprase function
        // ((CheckBoxPreference)preference).setChecked(false);
        Intent intent = new Intent(getActivity(), KeyCachingService.class);
        intent.setAction(KeyCachingService.DISABLE_ACTION);
        getActivity().startService(intent);
```

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 8]
```diff
@@ -119,6 +141,11 @@ private void initializeVisibility() {
         findPreference(TextSecurePreferences.SCREEN_LOCK).setEnabled(false);
       }
     } else {
+      findPreference("pref_enable_passphrase_temporary").setVisible(true); // JW
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: stale comment

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 9]
```diff
@@ -87,6 +94,13 @@ public void onResume() {
     else                                                         initializeScreenLockTimeoutSummary();
 
     disablePassphrase.setChecked(!TextSecurePreferences.isPasswordDisabled(getActivity()));
+    // Set the choice between passphrase and Android screenlock. Because this preference
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: better to insert a new line before long comments

**Comment by valldrac** [comment]
Great PR @johanw666. It's pretty neat. To test it in the upcoming x3, please target branch "x3"

File: src/org/thoughtcrime/securesms/util/TextSecurePreferences.java (hunk 1 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: Sorry, where? The one before "public"? All those constants there have a leading whitespace.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 2 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: OK, I'll change that.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 3 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: Yes I know but this is the original Signal code. I can make it into a constant if you like.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 4 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: Oops... Leftover from my fork where I use this tag to find my changes in the code. I'll remove it.

File: src/org/thoughtcrime/securesms/util/TextSecurePreferences.java (hunk 5 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: OK.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 7 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: I'll look at this later.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 8 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: Same oops.

File: src/org/thoughtcrime/securesms/util/TextSecurePreferences.java (hunk 1 above)
Comment (valldrac) [reply src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: The empty line at 187, one space before the new line char. Don't know how to see it in github, but git diff mark will it as red.

Comment (johanw666) [reply src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: OK, I had not noticed that. Probably an issue with Notepad++. I'll remove it.

Comment (johanw666) [reply src/org/thoughtcrime/securesms/util/TextSecurePreferences.java]: Fuxed in d48f838

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java (hunk 8 above)
Comment (johanw666) [reply src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: Fixed in d48f838

**Comment by valldrac** [comment]
Thanks. It's perfect to me now. Can you please resolve the conflict with latest commit? I merged current stable version 4.44.7 into x3. Just accept both changes from softlock/screenlock-distinct and master.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 10]
```diff
@@ -306,6 +331,59 @@ public boolean onPreferenceChange(final Preference preference, Object newValue)
...
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+      boolean screenlockEnabled = (boolean) newValue;
+
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        TextSecurePreferences.setScreenLockEnabled(getContext(), screenlockEnabled);
+        enableScreenLock.setChecked(screenlockEnabled);
+      }
+      else {
+        if (screenlockEnabled) {
```
Comment (valldrac) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: } else if (screenlockEnabled) {

**Pull request merged**